scipy/optimize/_tstutils.py

```diff
@@ -1,61 +1,678 @@
+     f4 has a left- and right- discontinuity at the root
+     f5 has a root at 1 replacing a 1st order pole
+     f6 is randomly positive on one side of the root,
+     randomly negative on the other.
+     f4 - f6 are not continuous at the root.
+
+  'aps': The test problems in the 1995 paper
+     TOMS "Algorithm 748: Enclosing Zeros of Continuous Functions"
+     by Alefeld, Potra and Shi.  Real-valued functions of
+     real-valued inputs on an interval with a zero.
+     Suitable for methods which start with an enclosing interval, and
+     derivatives up to 2nd order.
+
+  'complex': Some complex-valued functions of complex-valued inputs.
+     No enclosing bracket is provided.
+     Suitable for methods which use one or more starting values, and
+     derivatives up to 2nd order.
+
+  The test cases are provided as a list of dictionaries. The dictionary
+  keys will be a subset of:
+  ["f", "fprime", "fprime2", "args", "bracket", "smoothness",
+  "a", "b", "x0", "x1", "root", "ID"]
+"""
+
+# Sources:
+#  [1] Alefeld, G. E. and Potra, F. A. and Shi, Yixun,
+#      "Algorithm 748: Enclosing Zeros of Continuous Functions",
+#      ACM Trans. Math. Softw. Volume 221(1995)
+#       doi = {10.1145/210089.210111},
+
 from random import random
 
+import numpy as np
+
 from scipy.optimize import zeros as cc
 
+# "description" refers to the original functions
+description = """
+f2 is a symmetric parabola, x**2 - 1
+f3 is a quartic polynomial with large hump in interval
+f4 is step function with a discontinuity at 1
+f5 is a hyperbola with vertical asymptote at 1
+f6 has random values positive to left of 1, negative to right
+
+of course these are not real problems. They just test how the
+'good' solvers behave in bad circumstances where bisection is
+really the best. A good solver should not be much worse than
+bisection in such circumstance, while being faster for smooth
+monotone sorts of functions.
+"""
+
 
 def f1(x):
-    return x*(x-1.)
+    r"""f1 is a quadratic with roots at 0 and 1"""
+    return x * (x - 1.)
+
+
+def f1_fp(x):
+    return 2 * x - 1
+
+
+def f1_fpp(x):
+    return 2
 
 
 def f2(x):
+    r"""f2 is a symmetric parabola, x**2 - 1"""
     return x**2 - 1
 
 
+def f2_fp(x):
+    return 2 * x
+
+
+def f2_fpp(x):
+    return 2
+
+
 def f3(x):
-    return x*(x-1.)*(x-2.)*(x-3.)
+    r"""A quartic with roots at 0, 1, 2 and 3"""
+    return x * (x - 1.) * (x - 2.) * (x - 3.)  # x**4 - 6x**3 + 11x**2 - 6x
+
+
+def f3_fp(x):
+    return 4 * x**3 - 18 * x**2 + 22 * x - 6
+
+
+def f3_fpp(x):
+    return 12 * x**2 - 36 * x + 22
 
 
 def f4(x):
+    r"""Piecewise linear, left and right discontinuous at x=1, the root."""
     if x > 1:
-        return 1.0 + .1*x
+        return 1.0 + .1 * x
     if x < 1:
-        return -1.0 + .1*x
+        return -1.0 + .1 * x
     return 0
 
 
 def f5(x):
+    r"""Hyperbola with a pole at x=1, but pole replaced with 0.  Not continuous at root."""
```

Comment:
A bit late, but better late than never: this line (and many more below) exceeds 79 characters.  Please follow [PEP 8](https://www.python.org/dev/peps/pep-0008/) in the future.

Reply:
@WarrenWeckesser I switched to pycodestyle as that is what appears to be used in the first of the TravicCI builds.  pycodestyle  did not (and still isn't) showing me any warnings about lines over 79 lines long.  But just now when I invoke pep8 directly these warnings suddenly appear.  Is there some configuration that needs to be applied to pycodestyle?

Reply:
This is error code `E501`, which is ignored in our `tox.ini` and `pycodestyle` respects that. When you copy `zeros.py` outside of the source tree and run `pycodestyle` over it, you do get ~15 "line too long" warnings.

We ignore it because we don't want to fix it for the whole code base at once (messes up git history and causes merge conflicts). There's no good way around that really. I just set up my editor to break lines at 79 chars.

Reply:
There is a lot of older code in SciPy that does not comply with PEP 8, so in the file `tox.ini`, `pycodestyle` is configured to ignore "line too long" errors (`E501`).  This means we have to rely on contributors and reviewers to be familiar with the PEP 8 guidelines to avoid adding *new* code that doesn't follow PEP 8.   Sometimes even experienced SciPy developers forget to check. :)   To check my own modifications when working on SciPy, sometimes I temporarily rename `tox.ini` and run `pycodestyle` to make sure I haven't added any code that doesn't follow the guidelines.

Ignoring `E501` also means that cases where we want to allow longer lines won't get flagged as style errors.  For example, the big table that you added at the beginning of `__init__.py` might be the best option for presenting that information.  In that case, violating the guideline for the maximum line length is--perhaps!--justified.  (But it does appear to me that some of the columns in that table could be shortened just by using two lines for the headings.  I just took a look at the table in an ipython session, and I had to make my terminal wider to read it.)